Source/WebCore/loader/ResourceMonitor.cpp

```diff
@@ -0,0 +1,185 @@
+}
+
+void ResourceMonitor::setEligibility(Eligibility eligibility)
+{
+    if (m_eligibility == eligibility || m_eligibility == Eligibility::Eligible)
+        return;
+
+    m_eligibility = eligibility;
+
+    if (RefPtr parentMonitor = parentResourceMonitor())
+        parentMonitor->setEligibility(eligibility);
+    else
+        checkCondition();
+}
+
+void ResourceMonitor::setDocument(URL&& url)
+{
+    m_frameURL = WTFMove(url);
+    didReceiveResource(m_frameURL, ContentExtensions::ResourceType::Document);
+}
+
+void ResourceMonitor::didReceiveResource(const URL& url, ResourceTypeSet resourceType)
+{
+    if (m_eligibility == Eligibility::Eligible)
+        return;
+
+    RefPtr page = protectedFrame()->mainFrame().page();
+    if (!page)
+        return;
+
+    ResourceLoadInfo info = {
+        .resourceURL = url,
+        .mainDocumentURL = page->mainFrameURL(),
+        .frameURL = m_frameURL,
+        .type = resourceType
+    };
+
+    Checker::singleton().checkEligibility(info, [weakThis = WeakPtr(this)](Eligibility eligibility) mutable {
+        callOnMainRunLoop([weakThis = WTFMove(weakThis), eligibility] {
+            if (RefPtr protectedThis = weakThis.get())
+                protectedThis->setEligibility(eligibility);
+        });
+    });
+}
+
+void ResourceMonitor::addNetworkUsage(size_t bytes)
+{
+    m_networkUsage += bytes;
+
+    if (RefPtr parentMonitor = parentResourceMonitor())
+        parentMonitor->addNetworkUsage(bytes);
+    else
+        checkCondition();
+}
+
+void ResourceMonitor::checkCondition()
+{
+    ASSERT(!parentResourceMonitor());
+    if (m_eligibility != Eligibility::Eligible)
+        return;
+
+    if (!m_networkUsageExceed && m_networkUsage >= 4 * 1024 * 1024) {
+        m_networkUsageExceed = true;
+        protectedFrame()->networkUsageDidExceedThreshold();
+    }
+}
+
+Ref<LocalFrame> ResourceMonitor::protectedFrame() const
+{
+    return m_frame.get();
+}
+
+ResourceMonitor* ResourceMonitor::parentResourceMonitor() const
+{
+    return protectedFrame()->parentResourceMonitor();
+}
+
+ResourceMonitor::Checker& ResourceMonitor::Checker::singleton()
+{
+    static NeverDestroyed<ResourceMonitor::Checker> globalChecker;
+    return globalChecker;
+}
+
+ResourceMonitor::Checker::Checker()
+    : m_workQueue { WorkQueue::create("ResourceMonitor::Checker Work Queue"_s) }
+{
+}
+
+static ResourceMonitor::ResourceLoadInfo isolatedCopy(const ResourceMonitor::ResourceLoadInfo& info)
+{
+    return {
+        .resourceURL = info.resourceURL.isolatedCopy(),
+        .mainDocumentURL = info.mainDocumentURL.isolatedCopy(),
+        .frameURL = info.frameURL.isolatedCopy(),
+        .type = info.type,
+        .mainFrameContext = info.mainFrameContext,
+    };
+}
+
+void ResourceMonitor::Checker::checkEligibility(const ResourceLoadInfo& info, QueryCallback&& callback)
+{
+    m_workQueue->dispatch([this, info = isolatedCopy(info), callback = WTFMove(callback)] {
```

Comment:
Since this class uses multiple thread, please add `ASSERT(isMainThread());` or `ASSERT(!isMainThread());` at the beginning of every member function to document what thread they're meant to be called on.